File: 프로그래머스/Level 2/모음 사전.cpp
#include <string>
#include <vector>

using namespace std;

char ch[5] = {'A', 'E', 'I', 'O', 'U'};
int answer = 0;
int cnt = 0;

void dfs(string str, string target) {
    if (str.compare(target) == 0) {
        answer = cnt;
        return ;
    }
    
    cnt++;
    
    if (str.size() >= 5)
        return ;
    
    for (int i = 0; i < 5; i++) {
        dfs(str + ch[i], target);
    }
}

int solution(string word) {
    
    dfs("", word);
    
    return answer;
}

File: 프로그래머스/Level 2/모음 사전.java
import java.util.*;

class Solution {
    String[] alpha = { "A", "E", "I", "O", "U" };
    ArrayList<String> list = new ArrayList<>();
    
    public void dfs(String str, int len) {
        list.add(str);
        
        if (len == 5)
            return;
        
        for (int i = 0; i < 5; i++)
            dfs(str + alpha[i], len + 1);
    }
    
    public int solution(String word) {
        int answer = 0;
        
        dfs("", 0);
        
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).equals(word))
                return i;
        }
        return answer;
    }
}
